pkg/controller/v1alpha2/core/scopes/healthscope/healthscope.go

```diff
@@ -205,12 +232,55 @@ func CheckDaemonsetHealth(ctx context.Context, client client.Client, ref runtime
 		r.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
 		return r
 	}
-	r.Target.UID = daemonset.GetUID()
+	r.ComponentName = getComponentNameFromLabel(&daemonset)
+	r.TargetWorkload.UID = daemonset.GetUID()
+	r.Diagnosis = fmt.Sprintf(infoFmtReady, daemonset.Status.NumberReady, daemonset.Status.DesiredNumberScheduled)
 
+	// Health criteria
 	if daemonset.Status.NumberUnavailable != 0 {
-		r.Diagnosis = fmt.Sprintf(errFmtResourceNotReady, daemonset.Status)
 		return r
 	}
-	r.IsHealthy = true
+	r.HealthStatus = StatusHealthy
 	return r
 }
+
+// CheckByHealthCheckTrait checks health condition through HealthCheckTrait.
+func CheckByHealthCheckTrait(ctx context.Context, c client.Client, wlRef runtimev1alpha1.TypedReference, ns string) *WorkloadHealthCondition {
+	//TODO(roywang) implement HealthCheckTrait feature
+	return nil
+}
+
+// CheckUnknownWorkload handles unknown type workloads.
+func CheckUnknownWorkload(ctx context.Context, c client.Client, wlRef runtimev1alpha1.TypedReference, ns string) *WorkloadHealthCondition {
+	healthCondition := &WorkloadHealthCondition{
+		TargetWorkload: wlRef,
+		HealthStatus:   StatusUnknown,
+		Diagnosis:      fmt.Sprintf(infoFmtUnknownWorkload, wlRef.APIVersion, wlRef.Kind),
+	}
+
+	wl := &unstructured.Unstructured{}
+	wl.SetGroupVersionKind(wlRef.GroupVersionKind())
+	if err := c.Get(ctx, client.ObjectKey{Namespace: ns, Name: wlRef.Name}, wl); err != nil {
+		healthCondition.Diagnosis = errors.Wrap(err, errHealthCheck).Error()
+		return healthCondition
+	}
+	healthCondition.ComponentName = wl.GetLabels()[oam.LabelAppComponent]
+
+	// for unknown workloads, just show status instead of precise diagnosis
+	wlStatus, _, _ := unstructured.NestedMap(wl.UnstructuredContent(), "status")
```

Comment:
consider the CR didn't have status field? What will happen here?

Reply:
for such CR, wlStatus is nil returned by `NuestedMap` and `json.Marshal(nil)` will return `"null",nil`. So `healthCondition.WorkloadStatus = "null"`

Reply:
maybe I should use a specific explicit warning, e.g. "cannot get status of the unknown workload", instead of "null"

Reply:
null is ok